modules/afmongodb/afmongodb-legacy-uri.c

```diff
@@ -0,0 +1,301 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "afmongodb-legacy-uri.h"
+#include "afmongodb-legacy-private.h"
+
+#define DEFAULTHOST "127.0.0.1"
+#define SOCKET_TIMEOUT_FOR_MONGO_CONNECTION_IN_MILLISECS 60000
+
+static gboolean
+_parse_addr(const char *str, char **host, gint *port)
+{
+  if (!host || !port)
+    {
+      msg_debug("Host or port reference should not be NULL");
+      return FALSE;
+    }
+  char *proto_str = g_strdup_printf("mongodb://%s", str);
+  mongoc_uri_t *uri = mongoc_uri_new(proto_str);
+  if (!uri)
+    {
+      msg_error("Cannot parse MongoDB URI", evt_tag_str("uri", proto_str));
+      g_free(proto_str);
+      return FALSE;
+    }
+
+  const mongoc_host_list_t *hosts = mongoc_uri_get_hosts(uri);
+  if (!hosts || hosts->next)
+    {
+      if (hosts)
+        msg_error("Multiple hosts found in MongoDB URI", evt_tag_str("uri", proto_str));
+      else
+        msg_error("No host found in MongoDB URI", evt_tag_str("uri", proto_str));
+      g_free(proto_str);
+      mongoc_uri_destroy(uri);
+      return FALSE;
+    }
+  *port = hosts->port;
+  *host = g_strdup(hosts->host);
+  mongoc_uri_destroy(uri);
+  if (!*host)
+    {
+      msg_error("NULL hostname", evt_tag_str("uri", proto_str));
+      g_free(proto_str);
+      return FALSE;
+    }
+  g_free(proto_str);
+  return TRUE;
+}
+
+static void
+_free_host_port(gpointer data)
+{
+  MongoDBHostPort *hp = (MongoDBHostPort *)data;
+  g_free(hp->host);
+  hp->host = NULL;
+  g_free(hp);
+}
+
+static gboolean
+_append_host(GList **list, const char *host, gint port)
+{
+  if (!list)
+    return FALSE;
+  MongoDBHostPort *hp = g_new0(MongoDBHostPort, 1);
+  hp->host = g_strdup(host);
+  hp->port = port;
+  *list = g_list_append(*list, hp);
+  return TRUE;
+}
+
+static gboolean
+_append_legacy_servers(MongoDBDestDriver *self)
+{
+  if (self->port != MONGO_CONN_LOCAL)
+    {
+      if (self->address || self->port)
+        {
+          gchar *host = NULL;
+          gint port = 0;
+          if (!_parse_addr(self->address, &host, &port) || (!host))
+            {
+              msg_error("Cannot parse the primary host",
+                        evt_tag_str("primary", self->address),
+                        evt_tag_str("driver", self->super.super.super.id));
+              return FALSE;
+            }
+          g_free(host);
+
+          port = self->port ? self->port : MONGOC_DEFAULT_PORT;
+          const gchar *address = self->address ? self->address : DEFAULTHOST;
+          gchar *srv = g_strdup_printf("%s:%d", address, port);
+          self->servers = g_list_prepend(self->servers, srv);
+          g_free(self->address);
+          self->address = NULL;
```

Comment:
It bothers me a little that we patch address into servers in a function that should only use the configured data as read only.
This append legacy servers looks a bit hairy